specs/phase1/beacon-chain.md

```diff
@@ -1057,51 +485,124 @@ def process_epoch(state: BeaconState) -> None:
     process_reveal_deadlines(state)
     process_challenge_deadlines(state)
     process_slashings(state)
+    process_pending_headers(state)
+    charge_confirmed_header_fees(state)
+    reset_pending_headers(state)
     process_final_updates(state)  # phase 0 final updates
     process_phase_1_final_updates(state)
 ```
 
+#### Pending headers
+
+```python
+
+def process_pending_headers(state: BeaconState):
+    for slot in range(SLOTS_PER_EPOCH):
+        for shard in range(SHARD_COUNT):
+            # Pending headers for this (slot, shard) combo
+            candidates = [
+                c for c in state.previous_epoch_pending_shard_headers if
+                (c.slot, c.shard) == (slot, shard)
+            ]
+            if True not in [c.confirmed for c in candidates]:
+                # The entire committee (and its balance)
+                full_committee = get_beacon_committee(state, slot, shard)
+                full_committee_balance = get_total_balance(state, full_committee)
+                # The set of voters who voted for each header
+                # (and their total balances)
+                voting_sets = [
+                    [v for i, v in enumerate(full_committee) if c.votes[i]]
+                    for c in candidates
+                ]
+                voting_balances = [
+                    get_total_balance(state, voters)
+                    for voters in voting_sets
+                ]
+                # Get the index with the most total balance voting for them.
+                # NOTE: if two choices get exactly the same voting balance,
+                # the candidate earlier in the list wins
+                if max(voting_balances) > 0:
+                    winning_index = voting_balances.index(max(voting_balances))
+                else:
+                    # If no votes, zero wins
+                    winning_index = [c.root for c in candidates].index(Root())
+                candidates[winning_index].confirmed = True
+    for slot in range(SLOTS_PER_EPOCH):
+        for shard in range(SHARD_COUNT):
+            state.most_recent_confirmed_commitments[shard][slot] = DataCommitment()
+    for c in state.previous_epoch_pending_shard_headers:
+        if c.confirmed:
+            state.most_recent_confirmed_commitments[c.shard][c.slot % SLOTS_PER_EPOCH] = c.commitment
+```            
+
+```python
+def charge_confirmed_header_fees(state: BeaconState) -> None:
+    new_gasprice = state.shard_gasprice
+    adjustment_quotient = get_active_shard_count(state) * SLOTS_PER_EPOCH * GASPRICE_ADJUSTMENT_COEFFICIENT
+    for slot in range(SLOTS_PER_EPOCH):
+        for shard in range(SHARD_COUNT):
+            confirmed_candidates = [
+                c for c in state.previous_epoch_pending_shard_headers if
+                (c.slot, c.shard, c.confirmed) == (slot, shard, True)
+            ]
+            if confirmed_candidates:
```

Comment:
Why should the proposer only be charged for confirmed blocks? If they are doing their job reasonably, then they should be able to get their proposal confirmed, so it would be fair to always charge them. After all, an unavailable block might cause as much, if not more work on attesters.

Reply:
Good question! Need to think through the consequences of this. The status quo in ethereum is certainly "you pay only if you get included", though I can see the arguments for changing it.

Reply:
I think the main consequence is that we introduce a griefing vector by a committee, to conspire to not confirm a shard block in order to charge a proposer the gas fee without a reward.

Second possible problem could be that under condition where you consider it less likely that a block will be confirmed, e.g. when latency is high, it would discourage block production. In fact this threshold could be quite low: if we expect that a block producer makes 10% profit, then any confirmation probability under 90% would discourage block production.

Reply:
Charging for unconfirmed blocks would make equivocation to fill up proposals more expensive (https://github.com/ethereum/eth2.0-specs/pull/2146#discussion_r550452269)